### ISSUE TITLE
Train tiny-yolov2 with VOC dataset

### ISSUE THREAD

**Issue by barry41409** [AlexeyAB/darknet#2198]
I want to train tiny-yolov2 with VOC dataset.
VOC image size is 416*416 and use ./cfg/tiny-yolo-voc.cfg.

Where can I find pre-trained weight for tiny-yolov2 with VOC dataset 
like below weight(darknet53.conv.74)?
(darknet.exe detector train data/voc.data cfg/yolov3-voc.cfg darknet53.conv.74)

I can't find  pre-trained weight, the accuracy of tiny-yolo is lower than paper(maybe 4X%)
Does anyone train tiny-yolo model only using .cfg file?
and the accuracy can achieve 57.1% from the https://pjreddie.com/darknet/yolov2/ .
How can I modify the cfg file?

**Comment by AlexeyAB** [comment]
@barry41409 Hi,

How to get pre-trained weights file for yolov2-tiny: https://github.com/AlexeyAB/darknet/tree/47c7af1cea5bbdedf1184963355e6418cb8b1b4f#how-to-train-tiny-yolo-to-detect-your-custom-objects

----

> I want to train tiny-yolov2 with VOC dataset.
> VOC image size is 416*416 and use ./cfg/tiny-yolo-voc.cfg.

VOC images are not 416x416. And you shouldn't resize it to the 416x416 manually.

**Comment by barry41409** [comment]
hi @AlexeyAB, 
if I want to train tiny-yolo model with VOC by myself, I need pretrained weight to train model.
    https://github.com/AlexeyAB/darknet/issues/1877
    This command creates weights file yolov2-tiny.conv.13 that is suitable only for Training, not for 
    Detection:
    1. 
        darknet partial cfg/yolov2-tiny-voc.cfg yolov2-tiny-voc.weights yolov2-tiny.conv.13 13
    2.
        darknet detector train data/my.data tiny-yolo-voc.cfg yolov2-tiny.conv.13
    After training, I can get same accuracy like paper and get new weight. 
    Can I use this cmd?

What does partial function do? What is different from yolov2-tiny-voc.weights and darknet19_448.conv.23?
The darknet19_448.conv.23 is from darknet and where is the yolov2-tiny-voc.weights from?

**Comment by AlexeyAB** [comment]
@barry41409 Hi,

`partial` saves only the first several layers

* yolov2-tiny-voc.weights - is **Detector** (Yolo v2 tiny) that says what object are on the image **and where**

* darknet19_448.weights - is Classifier (Darknet) that says what objects are on the image

* darknet19_448.conv.23 - **first 23 layers** of Classifier (Darknet) that **are the same as in Detector** (Yolo v2) not tiny

----

> The darknet19_448.conv.23 is from darknet and where is the yolov2-tiny-voc.weights from?

yolov2-tiny-voc.weights from the site http://pjreddie.com/media/files/yolov2-tiny-voc.weights ) It is trained on Pascal VOC dataset to detect objects.
While darknet19_448.weights is trained on ImageNet to classify images.

**Comment by barry41409** [comment]
hi @AlexeyAB 
thank you!
Now, I use yolov2-tiny.conv.13 with VOC but the mAP is 50.71%.
    ./darknet partial yolov2-tiny-voc.cfg yolov2-tiny-voc.weights yolov2-tiny.conv.13 13
    ./darknet detector train voc.data yolov2-tiny-voc_train.cfg yolov2-tiny.conv.13
    ./darknet detector map voc.data yolov2-tiny-voc_test.cfg backup/yolov2-tiny-voc_train_final.weights

If I reload yolov2-tiny-voc.weights for testing, the mAP is 55.57%.
    ./darknet detector map voc.data yolov2-tiny-voc_test.cfg yolov2-tiny-voc.weight

Why the mAP are so low? The website of YOLOv2 is 57.1%(mAP).
Do I modify command or steps?

If I don't use yolov2-tiny.conv.13 for VOC, the mAP is too low.
Do I must use yolov2-tiny.conv.13 for training? 
    ./darknet detector train *.data yolov2-tiny-voc_train.cfg yolov2-tiny.conv.13
Do you know how to generate yolov2-tiny-voc.weights?

**Comment by AlexeyAB** [comment]
@barry41409 

* How many iterations did you train?
* What params did you use in the Makefile?
* What GPU did you use?
* Did you change anything in the cfg-file? What batch and subdivisions did you use?
* Did you use both VOC 2007+ VOC 2012 for training?

**Comment by barry41409** [comment]
Hi, @AlexeyAB, thank you!

- How many iterations did you train?
    40200

    [net]
    //Testing
    //batch=1
    //subdivisions=1
    //Training
    batch=64
    subdivisions=4
    width=416
    height=416
    channels=3
    momentum=0.9
    decay=0.0005
    angle=0
    saturation = 1.5
    exposure = 1.5
    hue=.1
    learning_rate=0.001
    max_batches = 40200
    policy=steps
    steps=-1,100,20000,30000
    scales=.1,10,.1,.1

- What params did you use in the Makefile?
    GPU=1
    CUDNN=1
    CUDNN_HALF=0
    OPENCV=1
    AVX=1
    OPENMP=1
    LIBSO=0

- What GPU did you use?
    GeForce GTX 1070

- Did you change anything in the cfg-file? What batch and subdivisions did you use?
    I don't change the cfg file. 
    For training, batch=64, subdivisions=4 -> save yolov2-tiny-voc_train.cfg
    For testing, batch=1, subdivisions=1 -> yolov2-tiny-voc_test.cfg

- Did you use both VOC 2007+ VOC 2012 for training?
    I only use VOC2007 for training.
    training result
    https://imgur.com/a/PmuiNpP

Do I must load yolov2-tiny.conv.13 for training with other dataset?
./darknet detector train *.data yolov2-tiny-voc_train.cfg yolov2-tiny.conv.13
Do you know how to generate yolov2-tiny-voc.weights?

**Comment by AlexeyAB** [comment]
> Do you know how to generate yolov2-tiny-voc.weights?

You can download it: https://pjreddie.com/media/files/yolov2-tiny-voc.weights

Try to train by using `batch=256 subdivisions=16`

**Comment by barry41409** [comment]
hi @AlexeyAB 
the weight from https://pjreddie.com/media/files/yolov2-tiny-voc.weights 
is different the website(https://pjreddie.com/darknet/yolov2/)?

Do I use partial cmd with https://pjreddie.com/media/files/yolov2-tiny-voc.weights 
or use the weight(https://pjreddie.com/media/files/yolov2-tiny-voc.weights) directly for training?

**Comment by AlexeyAB** [comment]
@barry41409 

> the weight from https://pjreddie.com/media/files/yolov2-tiny-voc.weights
> is different from the website(https://pjreddie.com/darknet/yolov2/)?

These files are the same.

to get ` yolov2-tiny.conv.13` use this command:

`./darknet partial cfg/yolov2-tiny-voc.cfg yolov2-tiny-voc.weights yolov2-tiny.conv.13 13`

**Comment by barry41409** [comment]
Thank you @AlexeyAB 
I use the yolov2-tiny.conv.13 and modify cfg as bellow.
    batch=256 
    subdivisions=16
    max_batches = 10200
    policy=steps
    steps=-1,25,5000,6000
    scales=.1,10,.1,.1
After training, I can get the mAP as same as the website 
and the yolov2-tiny-voc_final.weights as same as yolov2-tiny-voc.weights.

If I want to rotate image or do some image processing by myself,
I must load yolov2-tiny.conv.13, cfg and train the model again.
Which cfg do I use? the original or this new one.
I know darknet can't support rotation.

**Comment by barry41409** [comment]
Hi @AlexeyAB ,
Do I provide any information for training?
Do you have any suggestion for this problem?
Thanks for your help.

**Comment by AlexeyAB** [comment]
@barry41409 Hi,

> for thresh = 0.25, precision = -nan, recall = 0.00, F1-score = -nan
> for thresh = 0.25, TP = 0, FP = 0, FN = 12032, average IoU = 0.00 %
> 
> mean average precision (mAP) = 0.000000, or 0.00 %


It looks like your Training or Validation dataset is wrong.
Try to run and show the result
`darknet.exe detector calc_anchors data/voc.data -width 416 -height 416`

Then set `train=valid.txt` and run this command again and show me result.

**Comment by barry41409** [comment]
hi, @AlexeyAB 
Thank you.
I run the cmd as below.
./darknet detector calc_anchors my/voc.data -width 416 -height 416
 ```
num_of_clusters = 5, width = 416, height = 416 
 read labels from 2501 images 
 loaded 	 image: 2501 	 box: 6301
 all loaded. 

 calculating k-means++ ...

 avg IoU = 61.74 % 

Saving anchors to the file: anchors.txt 
anchors =  39, 71,  89,147, 139,287, 245,170, 318,336
```
then set the train=valid in voc.data.
./darknet detector calc_anchors my/voc.data -width 416 -height 416
```
 num_of_clusters = 5, width = 416, height = 416 
 read labels from 4952 images 
 loaded 	 image: 4952 	 box: 12032
 all loaded. 

 calculating k-means++ ...


 avg IoU = 62.18 % 

Saving anchors to the file: anchors.txt 
anchors =  39, 71,  91,149, 141,288, 252,175, 330,337

```

**Comment by AlexeyAB** [comment]
@barry41409 Your labels of Training and Validations datasets are correct.

* Do you use the latest version of this repository?
* Can you rename `yolov2-tiny-voc_train_new.cfg` to `yolov2-tiny-voc_train_new.txt` and attach it to your message?

Try to train from the begining by using such command:
`./darknet detector train source_jhan/voc.data source_jhan/yolov2-tiny-voc_train_new.cfg source_jhan/yolov2-tiny.conv.13 -map`

* And see at the Loss/mAP-chart, what Loss and mAP will be after 4000 iterations?

**Comment by barry41409** [comment]
hi @AlexeyAB ,thank you.
I use latest version, Does it renew recently?
Here is my cfg data, [link](https://drive.google.com/file/d/1MpfrGMxzl4elRnr6Jggp2zu5-FArgv9o/view?usp=sharing)

I am running the cmd 
./darknet detector train source_jhan/voc.data source_jhan/yolov2-tiny-voc_train_new.cfg source_jhan/yolov2-tiny.conv.13 -map

**Comment by AlexeyAB** [comment]
@barry41409 It was renewed 4 hours ago: https://github.com/AlexeyAB/darknet/commits/master

> I am running the cmd
> ./darknet detector train source_jhan/voc.data source_jhan/yolov2-tiny-voc_train_new.cfg source_jhan/yolov2-tiny.conv.13 -map

So after each 100 iterations will be created `chart.png` image.
After 4000 iterations drag-n-drop it to your message.

----

Also try to train with default
```
max_batches = 40200
policy=steps
steps=-1,100,20000,30000
scales=.1,10,.1,.1
```

**Comment by barry41409** [comment]
hi @AlexeyAB ,
1. use [yolov2-tiny-voc_train_new.cfg](https://ufile.io/b1vc7)
```
batch=256
subdivisions=16
max_batches = 10200
policy=steps
steps=-1,25,5000,6000
scales=.1,10,.1,.1
```
./darknet detector train source_jhan/voc.data source_jhan/yolov2-tiny-voc_train_new.cfg source_jhan/yolov2-tiny.conv.13 -map
[result](https://imgur.com/a/JCSOt9r)
(sorry, the result was break because the memory error
I reload the last weight and continue training.)

2. use [my cfg](https://ufile.io/g0u3k)
```
batch=64
subdivisions=4
max_batches = 40200
policy=steps
steps=-1,100,20000,30000
scales=.1,10,.1,.1
```
./darknet detector train source_jhan/voc.data source_jhan/yolov2-tiny-voc_train.cfg source_jhan/yolov2-tiny.conv.13 -map
[result](https://imgur.com/a/GDspmmW)

I find subdivisions=4 in my cfg different from subdivisions=2 in original cfg.
Does subdivisions effect the result?
Have you trained tiny-yolov2?
We can cross validation for the cfg file.

**Comment by AlexeyAB** [comment]
@barry41409 Hi,

> Does subdivisions effect the result?

Almost no.

> Have you trained tiny-yolov2?

For my own datasets.

**Comment by barry41409** [comment]
@AlexeyAB  Thanks, 
Dose my result,cfg correct?
How many classes are in your dataset and how many images are in valid and train? 
Do you have any idea or suggestion about training tiny-yolov2 with voc2007&2012?

I have problem about partial, 
If I use same dataset VOC2007&2012 and same class in cfg,
Why do I partial 13 layer pretrained weight for tiny yolov2?

`yolov2-tiny-voc.weights - is Detector (Yolo v2 tiny) that says what object are on the image and where`
Training with yolov2-tiny-voc.weights, the result is lower than the website(mAP: 57.1%).
Is it normal?

**Comment by AlexeyAB** [comment]
@barry41409 
> Why do I partial 13 layer pretrained weight for tiny yolov2?

Because remaining layers can depend on classes that can be different between default `yolov2-tiny-voc.weights ` and your.

> Training with yolov2-tiny-voc.weights, the result is lower than the website(mAP: 57.1%).
> Is it normal?

What mAP can you get?

**Comment by barry41409** [comment]
hi @AlexeyAB ,
testing with original yolov2-tiny-voc.weights 
```
./darknet detector map source_jhan/voc.data source_jhan/yolov2-tiny-voc_test.cfg yolov2-tiny-voc.weights 
mAP---->55.57%
```

training with yolov2-tiny-voc.weights, testing reload the final weight from training
```
./darknet detector train source_jhan/voc.data source_jhan/yolov2-tiny-voc_train.cfg yolov2-tiny-voc.weights 
./darknet detector map source_jhan/voc.data source_jhan/yolov2-tiny-voc_test.cfg backup/yolov2-tiny-voc_train_final.weights 
mAP---->56.49%
```

training with partial yolov2-tiny-voc.weights, testing reload the final weight from training
```
./darknet partial source_jhan/yolov2-tiny-voc.cfg yolov2-tiny-voc.weights yolov2-tiny.conv.13 13
./darknet detector train source_jhan/voc.data source_jhan/yolov2-tiny-voc_train.cfg yolov2-tiny.conv.13 13 
./darknet detector map source_jhan/voc.data source_jhan/yolov2-tiny-voc_test.cfg backup/yolov2-tiny-voc_train_final.weights 
mAP---->50.71%
```
Why does darknet start from 40101 iterations when training with yolov2-tiny-voc.weights? 

training without any weight, testing reload the final weight from training
```
./darknet detector train source_jhan/voc.data source_jhan/yolov2-tiny-voc_train.cfg
./darknet detector map source_jhan/voc.data source_jhan/yolov2-tiny-voc_test.cfg backup_noreload/yolov2-tiny-voc_train_final.weights
mAP---->30.8%
```

I restart training tiny yolov2 with VOC2007&2012(class is same with website),
which weight and cfg do I use? That can achieve the mAP like the website.
(How can I get the final weight(yolov2-tiny-voc.weights) by training tiny-yolov2?)

If I want to train custom dataset, I should use partial weight and modify cfg.
Do I modify training policy about steps, scales....?
